swupd: create packs with fullfiles

cmarcelo opened the pull request:
Add a function WritePack, that will take a pair of from/to versions of
a bundle, and produce a single tar with all the needed files to
perform an update of that bundle. For now, we are only packing
fullfiles -- which is functional but leads to larger pack sizes. Once
deltas are avaialble, we can make packs smaller.

The function supports an optional chrootDir argument, that will be
used to read the files directly instead of uncompressing them from the
"files/" directory.

Different from swupd-server, we don't copy the files to an staged
directory and then archive/compress. By taking advantage of Go
archive/tar package, we can create the tar file directly. Because of
this, the speed up from the chrootDir optimization is smaller in the
Go version.

Also add a test program create-pack, that has the equivalent
functionality of pack-maker.sh: it can generate a single pack for a
from/to pair, or all the relevant packs from a from/to pair. In the
second case, it will dig into the MoM to find the actual from/to pairs
for each relevant bundle.

The test program already contain some logic that will later be adopted
by Mixer and/or the swupd package.
    
Signed-off-by: Caio Marcelo de Oliveira Filho <caio.oliveira@intel.com>

cmarcelo committed:
swupd: create packs with fullfiles

Add a function WritePack, that will take a pair of from/to versions of
a bundle, and produce a single tar with all the needed files to
perform an update of that bundle. For now, we are only packing
fullfiles -- which is functional but leads to larger pack sizes. Once
deltas are avaialble, we can make packs smaller.

The function supports an optional chrootDir argument, that will be
used to read the files directly instead of uncompressing them from the
"files/" directory.

Different from swupd-server, we don't copy the files to an staged
directory and then archive/compress. By taking advantage of Go
archive/tar package, we can create the tar file directly. Because of
this, the speed up from the chrootDir optimization is smaller in the
Go version.

Also add a test program create-pack, that has the equivalent
functionality of pack-maker.sh: it can generate a single pack for a
from/to pair, or all the relevant packs from a from/to pair. In the
second case, it will dig into the MoM to find the actual from/to pairs
for each relevant bundle.

The test program already contain some logic that will later be adopted
by Mixer and/or the swupd package.

Signed-off-by: Caio Marcelo de Oliveira Filho <caio.oliveira@intel.com>

cmarcelo commented:
NOTE: This PR was done on top of #16, so GitHub will show all the commits from #16 here until it gets merged. The only new patch is the last one, about packs.

matthewrsj commented on swupd/cmd/create-pack/main.go:
```diff
@@ -0,0 +1,237 @@
+	return uint32(parsed)
+}
+
+// TODO: Move to manifest.go as an exported function.
```
Agree, this should go in manifest.go and used in manifest creation (CreateManifests)

cmarcelo replied:
This is part of #23.

matthewrsj replied:
This function can be removed from this PR now right?

cmarcelo replied:
Done. Was waiting #23 getting in.

matthewrsj commented on swupd/packs.go:
```diff
@@ -1,10 +1,296 @@
+	}
+	result := &compressedTarReader{}
+	switch {
+	case h[0] == 0x1F && h[1] == 0x8B:
```
maybe a comment describing what these bytes represent? You are just reading header information to determine what to uncompress with, right?

cmarcelo replied:
Correct. And I see now how it might be confusing. Added a comment and changed the code to make the intention more explicit.

matthewrsj reviewed (commented):
Looks good overall, but you are making a lot of assumptions of what the directory structure looks like underneath the state directory (for example, www/ being the output directory). This is fine for now, but since those subdirectories are configurable that should be respected here.

matthewrsj commented:
@cmarcelo did you want this to stay open until the delta file functionality is added?

cmarcelo commented:
> @cmarcelo did you want this to stay open until the delta file functionality is added?

Given that it produces valid output, I think is fine merging it without delta, and then I'll add delta support to it on top.

I'm fixing the points you made and will send a new version of this PR.

cmarcelo commented:
> but you are making a lot of assumptions of what the directory structure looks like underneath the state directory (for example, www/ being the output directory). 

Fixed WritePack to not construct the output dir, but take it as parameter instead of stateDir.

The test program is assuming the default directory layout, we can add flags to let it specify different directories if needed.

matthewrsj commented on swupd/cmd/create-pack/main.go:
```diff
@@ -0,0 +1,237 @@
+				// Bundle doesn't exist in new version, no pack needed.
+				continue
+			}
+			if e.to == f.Version {
```
Are you waiting until now to check if fromVersion and toVersion are the same (e.to is set to toVersionUint for every bundle)? If you did this check earlier you wouldn't have to check for every bundle. From == to isn't even a valid pack, so this should be a check much further up.

cmarcelo replied:
Added an early distro version check.

We still need this per bundle check here, since different Clear Linux versions might have a bundle in the same version, so we want to skip it for packing purposes.

matthewrsj replied:
Oh, right. f.Version is for the manifest represented in the MoM, which can definitely have a version==toVersion.

matthewrsj replied:
actually... the file `f` is from `fromMoM`, which is the `Manifest.MoM` in `fromVersion`, which should be older than `toVersion` (and `e.to==toVersion`), so this case should never be hit because `f.Version` can not be any newer than `fromVersion` (the version of `fromMoM` from whence it came). What am I missing here?

matthewrsj replied:
disregard again, misread the code.

matthewrsj commented on swupd/cmd/create-pack/main.go:
```diff
@@ -0,0 +1,237 @@
+			fmt.Println("Warnings during pack:")
+		}
+		fmt.Printf("  %s\n", w)
+		if i == len(info.Warnings)-1 {
```
Instead of checking here, why not just do the `fmt.Println()` after the loop exits?

cmarcelo replied:
I just want the extra newline if there were warnings.

cmarcelo replied:
Ended up adding a if block around the loop and the prints.

matthewrsj commented on swupd/packs.go:
```diff
@@ -14,11 +14,306 @@
+			entry.Reason = "already in from manifest"
+			continue
+		}
+		if done[f.Hash] {
```
What about two differently-named files with identical content?

cmarcelo replied:
If files /usr/bin/a and /usr/bin/b have the same hash HASH, we will pack one first into staged/HASH, and the second will not be packed with reason "hash already packed".

matthewrsj replied:
right, of course. Never mind.

matthewrsj commented on swupd/packs.go:
```diff
@@ -14,11 +14,306 @@
+	}
+
+	err = tw.Close()
+	if err != nil {
```
you already deferred a func that will overwrite `info` with `nil` when `err != nil` so you should be able to just `return info, err` right here.

cmarcelo replied:
That defer was useful when I had more cases of deferred functions causing error. Since there is only one now, moved the "info = nil" there.

matthewrsj commented on swupd/packs.go:
```diff
@@ -14,11 +14,306 @@
+
+// Compression algorithms have "magic" bytes in the beginning of the file to identify them.
+var (
+	gzipMagic  = []byte{0x1F, 0x8B}
```
ah, thanks for adding this.

matthewrsj commented on swupd/packs.go:
```diff
@@ -14,11 +14,306 @@
+		return fmt.Errorf("failed to read fullfile %s: %s", fullfilePath, err)
+	}
+	// TODO: Check for error when closing this.
+	defer fullfileReader.Close()
```
might as well check the error here the same way you do elsewhere.

cmarcelo replied:
Done.

matthewrsj reviewed (commented):
Just a few minor things.

cmarcelo commented:
Fixing another issue I've found during more testing, will send the revised PR after that.

cmarcelo committed:
swupd: no need to close the output pipe in externalReader

The pipe is already closed by the call to cmd.Wait. At this point, we
are guaranteed to have Read all that we want, either an error or a
valid tar file.

Signed-off-by: Caio Marcelo de Oliveira Filho <caio.oliveira@intel.com>

matthewrsj reviewed (approved):
lgtm
